evm/src/cpu/kernel/asm/curve_add.asm

```diff
@@ -0,0 +1,383 @@
+    DUP4
+    // stack: lambda, x1 - x2, N, x2, lambda, x0, y0, x1, y1, retdest
+    MULMOD
+    // stack: lambda * (x1 - x2), x2, lambda, x0, y0, x1, y1, retdest
+    DUP7
+    // stack: y1, lambda * (x1 - x2), x2, lambda, x0, y0, x1, y1, retdest
+    SWAP1
+    // stack: lambda * (x1 - x2), y1, x2, lambda, x0, y0, x1, y1, retdest
+    %submod
+    // stack: y2, x2, lambda, x0, y0, x1, y1, retdest
+    SWAP5
+    // stack: x1, x2, lambda, x0, y0, y2, y1, retdest
+    POP
+    // stack: x2, lambda, x0, y0, y2, y1, retdest
+    SWAP5
+    // stack: y1, lambda, x0, y0, y2, x2, retdest
+    POP
+    // stack: lambda, x0, y0, y2, x2, retdest
+    POP
+    // stack: x0, y0, y2, x2, retdest
+    POP
+    // stack: y0, y2, x2, retdest
+    POP
+    // stack: y2, x2, retdest
+    SWAP2
+    // stack: retdest, x2, y2
+    JUMP
+
+// Assumption: (x0,y0) and (x1,y1) are valid points and x0 == x1
+ec_add_equal_first_coord:
+    JUMPDEST
+    // stack: x0, y0, x1, y1, retdest with x0 == x1
+    %bn_base
+    // stack: N, x0, y0, x1, y1, retdest
+    DUP3
+    // stack: y0, N, x0, y0, x1, y1, retdest
+    DUP6
+    // stack: y1, y0, N, x0, y0, x1, y1, retdest
+    ADDMOD
+    // stack: y1 + y0, x0, y0, x1, y1, retdest
+    PUSH ec_add_equal_points
+    // stack: ec_add_equal_points, y1 + y0, x0, y0, x1, y1, retdest
+    JUMPI
+    // stack: x0, y0, x1, y1, retdest
+    POP
+    // stack: y0, x1, y1, retdest
+    POP
+    // stack: x1, y1, retdest
+    PUSH ret_zero
+    // stack: ret_zero, x1, y1, retdest
+    JUMP
+
+
+// Assumption: x0 == x1 and y0 == y1
+ec_add_equal_points:
+    JUMPDEST
+    // stack: x0, y0, x1, y1, retdest
+    %bn_base
+    // stack: N, x0, y0, x1, y1, retdest
+    %bn_base
+    // stack: N, N, x0, y0, x1, y1, retdest
+    DUP3
+    // stack: x0, N, N, x0, y0, x1, y1, retdest
+    DUP1
+    // stack: x0, x0, N, N, x0, y0, x1, y1, retdest
+    MULMOD
+    // stack: x0^2, N, x0, y0, x1, y1, retdest with
+    PUSH 0x183227397098d014dc2822db40c0ac2ecbc0b548b438e5469e10460b6c3e7ea5 // 3/2 in the base field
+    // stack: 3/2, x0^2, N, x0, y0, x1, y1, retdest
+    MULMOD
+    // stack: 3/2 * x0^2, x0, y0, x1, y1, retdest
+    DUP3
+    // stack: y0, 3/2 * x0^2, x0, y0, x1, y1, retdest
+    %moddiv
+    // stack: lambda, x0, y0, x1, y1, retdest
+    PUSH ec_add_valid_points_with_lambda
+    // stack: ec_add_valid_points_with_lambda, lambda, x0, y0, x1, y1, retdest
+    JUMP
+
+// Assumption: (x0,y0) is a valid point.
+global ec_double:
+    JUMPDEST
+    // stack: x0, y0, retdest
+    DUP2
+    // stack: y0, x0, y0, retdest
+    DUP2
+    // stack: x0, y0, x0, y0, retdest
+    PUSH ec_add_equal_points
+    // stack: ec_add_equal_points, x0, y0, x0, y0, retdest
+    JUMP
+
+// Push the order of the BN254 base field.
+%macro bn_base
+    PUSH 0x30644e72e131a029b85045b68181585d97816a916871ca8d3c208c16d87cfd47
+%endmacro
+
+// Assumption: x, y < N and 2N < 2^256.
+// Note: Doesn't hold for Secp256k1 base field.
```

Comment:
Just thinking about secp for later, we could detect underflow as `x < x - y` (an idea Jakub used a lot in SIMD code), though the code gets a bit longer

```
%macro submod
    // stack: x, y
    SWAP1
    // stack: y, x
    DUP2
    // stack: x, y, x
    SUB
    // stack: x - y, x
    DUP1
    // stack: x - y, x - y, x
    SWAP2
    // stack: x, x - y, x - y
    LT
    // stack: x < x - y, x - y
    %secp_base
    // stack: N, x < x - y, x - y
    MUL
    // stack: N * (x < x - y), x - y
    ADD
    // (x-y) % N
%endmacro
```

If performance is an issue we can discuss adding a `SUBMOD` instruction, or some other feature like carry flags.